File: app/src/main/java/com/example/debtcalculator/services/DebtCalculatorService.kt
package com.example.debtcalculator.services

import com.example.debtcalculator.data.Balance
import com.example.debtcalculator.data.Group
import com.example.debtcalculator.data.Transaction
import com.example.debtcalculator.data.User
import kotlin.math.abs

class DebtCalculatorService {
    fun calculateBalances(group: Group): ArrayList<Balance> {
        val map = HashMap<User, Float>()
        for (expense in group.expenses) {
            val lender = expense.owner
            val pricePerUser = expense.amount / expense.payers.size

            // Update lender's balance
            val lenderBalance = if (expense.payers.contains(lender)) expense.amount - pricePerUser else expense.amount
            map.put(lender, (map.get(lender) ?: 0f) + lenderBalance)

            // Update payers balances
            for (payer in expense.payers) {
                if (lender != payer) map.put(payer, (map.get(payer) ?: 0f) - pricePerUser)
            }
        }
        val balances = ArrayList<Balance>()

        for (user in map.keys) {
            balances.add(Balance(user, (map.get(user) ?: 0f)))
        }
        return balances
    }

    fun determineTransactions(balances: ArrayList<Balance>): ArrayList<Transaction> {
        balances.sortBy { balance -> balance.balance }
        val transactions = ArrayList<Transaction>()
        var index = balances.size - 1
        for (balance in balances) {
            if (balance.balance >= 0) {
                break
            }
            // Keep sending money until balance is smaller than 1 cent/øre
            while (balance.balance < -0.01) {
                val receiver = balances.get(index)

                // Case for sender owing less or the same as the current receiver is missing
                if (abs(balance.balance) <= abs(receiver.balance)) {
                    transactions.add(Transaction(abs(balance.balance), balance.user, receiver.user))

                    // Set sender balance to 0 and add negative balance from sender to receiver
                    receiver.balance += balance.balance
                    balance.balance = 0f
                }
                // Case for sender owing more than current receiver is missing
                else {
                    transactions.add(
                        Transaction(
                            abs(balances.get(index).balance),
                            balance.user,
                            receiver.user
                        )
                    )
                    balance.balance += receiver.balance
                    receiver.balance = 0f
                    index--
                }
            }
        }
        return transactions
    }

    fun determineTransactions(group: Group): ArrayList<Transaction> {
        return determineTransactions(calculateBalances(group))
    }
}

File: backend/DebtCalculatorService.js
const Balance = require("./data/Balance");
const Transaction = require("./data/Transaction");


class DebtCalculatorService {

    calculateBalances(group) {
    const map = {}; 

    for (const expense of group.expenses) {
        const lender = expense.lender;
        const pricePerUser = expense.amount / expense.payers.length;

        if(expense.isPaid === 1) {
            continue;
        }

        // Check if lender is among payers
        const lenderOwesPart = expense.payers.some(p => p.phoneNumber === lender.phoneNumber);
        const lenderBalance = lenderOwesPart ? expense.amount - pricePerUser : expense.amount;

        map[lender.phoneNumber] = (map[lender.phoneNumber] || 0) + lenderBalance;

        for (const payer of expense.payers) {
            if (payer.phoneNumber !== lender.phoneNumber) {
                map[payer.phoneNumber] = (map[payer.phoneNumber] || 0) - pricePerUser;
            }
        }
    }

    const balances = [];
    for (const phone in map) {
        const user = group.users.find(u => u.phoneNumber === phone);
        balances.push({ user, balance: map[phone] });
    }

    return balances;
}


    determineTransactions(balances) {
        balances.sort((a, b) => a.balance - b.balance);

        let transactions = [];

        let index = balances.length -1;

        for(let balance of balances) {
            if(balance.balance >= 0) break;

            // Keep sending money until balance is smaller than 1 cent/øre
            while (balance.balance < -0.01) {
                let receiver = balances[index];

                // Case 1: Sender owes same amount or less than the receiver is owed
                if (Math.abs(balance.balance) <= Math.abs(receiver.balance)) {
                    transactions.push(new Transaction(Math.abs(balance.balance), balance.user, receiver.user));

                    // Set sender balance to 0 and add negative balance from sender to receiver
                    receiver.balance += balance.balance;
                    balance.balance = 0;

                // Case 2: Sender owes more than receiver is owed    
                } else {
                    transactions.push(new Transaction(Math.abs(receiver.balance), balance.user, receiver.user));

                    balance.balance += receiver.balance;
                    receiver.balance = 0;
                    index--;
                }
            }
        }
        return transactions;
    }

    determineTransactionsForgroup(group) {
        return this.determineTransactions(this.calculateBalances(group));
    }

}


module.exports = DebtCalculatorService;